src/NuGet.Core/NuGet.Packaging/Licenses/LicenseExpressionTokenizer.cs

```diff
@@ -0,0 +1,123 @@
+        /// </summary>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        internal bool HasValidCharacters()
+        {
+            var regex = new Regex("^[a-zA-Z0-9\\.\\-\\s\\+\\(\\)]+$");
+            return regex.IsMatch(_value);
+        }
+
+        /// <summary>
+        /// Given a string, tokenizes by space into operators and values. The considered operators are, AND, OR, WITH, (, and ). 
+        /// </summary>
+        /// <returns>Tokens</returns>
+        internal IEnumerable<LicenseExpressionToken> Tokenize()
+        {
+            var potentialTokens = _value.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (var token in potentialTokens)
+            {
+                var processingToken = token;
+
+                IList<LicenseExpressionToken> tokensAfterValue = null;
+
+                while (processingToken.Length > 0 && (processingToken[0] == '(' || processingToken[0] == ')'))
+                {
+                    yield return ParseBracket(processingToken[0]);
+
+                    processingToken = processingToken.Substring(1);
+                }
+
+                while (processingToken.Length > 0 && (processingToken[processingToken.Length - 1] == '(' || processingToken[processingToken.Length - 1] == ')'))
+                {
+                    if (tokensAfterValue == null)
+                    {
+                        tokensAfterValue = new List<LicenseExpressionToken>();
+
+                    }
+
+                    tokensAfterValue.Add(ParseBracket(processingToken[processingToken.Length - 1]));
+                    processingToken = processingToken.Substring(0, processingToken.Length - 1);
+                }
+
+                if (!string.IsNullOrEmpty(processingToken))
+                {
+                    yield return ParseTokenType(processingToken);
+                }
+
+                if (tokensAfterValue != null)
+                {
+                    foreach (var tokenAfterValue in tokensAfterValue)
+                    {
+                        yield return tokenAfterValue;
+                    }
+                }
+
+            }
+        }
+
+        private LicenseExpressionToken ParseBracket(char bracket)
+        {
+            if (bracket == '(')
+            {
+                return new LicenseExpressionToken(bracket.ToString(), LicenseTokenType.OPENING_BRACKET);
+            }
+            if (bracket == ')')
+            {
+                return new LicenseExpressionToken(bracket.ToString(), LicenseTokenType.CLOSING_BRACKET);
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Parses a token type given a string.
+        /// This method assumes that the brackets have been parsed out. 
+        /// </summary>
+        /// <param name="token"></param>
+        /// <returns>LicenseExpressionToken</returns>
```

Comment:
Add a more detailed description or remove. The type is already given by the function itself, no need to have it in the comments